### DISCUSSION TITLE
ligatures with hangul does not working correctly

### DISCUSSION THREAD

**Discussion by cloudchamb3r** [harfbuzz/harfbuzz#4853] [General]
I'm using [d2coding ligature font](https://github.com/user-attachments/files/16843174/d2codingliga.ttf.zip)
everything works fine, except for ligature with hangul. 

i found by enabling calt feature(disabled in `hb-ot-shaper-hangul.cc`) this issue can be resolved, but that makes some rendering issues on some old hangul characters with specific fonts, for example U+11A2 character with noto sans kor.

related problem is also addressed in this [vscode issue](https://github.com/microsoft/vscode/issues/89271)
this problem is not only for d2coding font, but also other fonts like [jetbrains mono with ligature](https://codepen.io/fpzailxb-the-encoder/pen/gONdXpb?editors=1100)

![liga ko](https://github.com/user-attachments/assets/05c3ae12-84db-4ba0-ac2d-fcd5b648ab85)
![liga jp](https://github.com/user-attachments/assets/540561f1-8445-4bca-a558-fa7115b009e1)
![liga en](https://github.com/user-attachments/assets/360bf121-cac4-4b02-9b82-cc74ffc03f0b)
![liga cn](https://github.com/user-attachments/assets/0b63dab1-e74a-44fe-85ea-9de91ed63b31)

**Comment by behdad** [comment]
How does it behave in other shapers (Windows, Mac)?

**Reply by cloudchamb3r** [reply]
It doesn't seem to be a OS or specific shaper issue. I'm having the same issue on Mac, Windows, and Android.

**Reply by khaledhosny** [reply]
You need to test in applications that do not use HarfBuzz on these platforms, so applications other than vscode which uses HarfBuzz everywhere. Try Notepad on Windows and TextEdit or Safari on macOS.

**Reply by cloudchamb3r** [reply]
In safari and iterm2 ligature works correctly. I'll test Windows later
## safari
<img width="1230" alt="Screenshot 2024-09-05 at 1 39 00 PM" src="https://github.com/user-attachments/assets/8611e0ee-3716-449c-8f05-6f3b35671417">

## iterm
<img width="221" alt="Screenshot 2024-09-05 at 1 38 26 PM" src="https://github.com/user-attachments/assets/1b6b7167-f9c4-4dd5-a94e-b8b00c84b67e">

## TextEdit
<img width="662" alt="Screenshot 2024-09-05 at 1 57 39 PM" src="https://github.com/user-attachments/assets/f9f308c8-fce3-4284-9b27-0b64257255b4">

## Notepad
![image](https://github.com/user-attachments/assets/88998710-9468-4ec5-a5be-0d1511bd99f1)

**Reply by behdad** [reply]
The comment in `hb-ot-shaper-hangul.cc` says:
```
static void 
override_features_hangul (hb_ot_shape_planner_t *plan) 
{ 
  /* Uniscribe does not apply 'calt' for Hangul, and certain fonts 
   * (Noto Sans CJK, Source Sans Han, etc) apply all of jamo lookups 
   * in calt, which is not desirable. */ 
  plan->map.disable_feature (HB_TAG('c','a','l','t')); 
} 
```
It's possible that 1. those fonts have been fixed, and 2. DWrite might have changed to apply calt in Hangul.

If we can answer those two questions, it would be easier to decide if we should remove that overrrid.

cc @jfkthame @nedley @PeterCon

**Reply by nedley** [reply]
This font doesn’t support the Hangul Jamo block so any examples using those characters will need fallback. ~~If I use a supported character like U+AC00 I also see the same problem in TextEdit.~~

~~Update: I must have been testing the wrong thing because now I can’t reproduce it.~~

Update 2: Suffice it to say that this will reproduce with CoreText in the general case.

**Reply by nedley** [reply]
Since the font doesn’t have Hangul script shaping data, perhaps that is affecting DWrite itemization?

**Reply by PeterConstable** [reply]
What kinds of character sequences are in question here? (I'm looking for text, not pictures of text.)

**Reply by PeterConstable** [reply]
DWrite's shaping library has a Hangul engine that applies 'locl', 'ccmp' and 'rclt' by default across a run, and applies 'ljmo', 'vjmo' and 'tjmo' on analysis elements of a jamo cluster.

**Reply by nedley** [reply]
The question here is which lookups are applied to the non-Hangul characters in text like `가=>`, where the last two characters are ligated by a 'calt' lookup.

**Reply by khaledhosny** [reply]
Would this be a text segmentation issue? Chrome and Firefox are classifying the common characters with the script of the Hangul characters next to them regardless of the font support. AFAIK script segmentation does not take font into account, would DirectWrite or CoreText be doing otherwise?

**Reply by nedley** [reply]
This is an issue for CoreText as well.

**Reply by PeterConstable** [reply]
In DWrite, the entire string is itemized as a Hangul run, so only 'locl', 'ccmp' and 'rclt' would get applied to the non-Hangul characters. ('calt' would be reasonable to expect, though.)

**Reply by behdad** [reply]
Is it time we enable `calt` in the Hangul shaper?
